integration/spark/src/main/spark3/java/io/openlineage/spark/agent/lifecycle/Spark3VisitorFactoryImpl.java

```diff
@@ -30,6 +30,10 @@
 
   public <D extends Dataset> List<PartialFunction<LogicalPlan, List<D>>> getCommonVisitors(
       OpenLineageContext context, DatasetFactory<D> factory) {
-    return super.getBaseCommonVisitors(context, factory);
+    DatasetFactory<OutputDataset> outputFactory = DatasetFactory.output(context.getOpenLineage());
+    return ImmutableList.<PartialFunction<LogicalPlan, List<D>>>builder()
+        .addAll(super.getBaseCommonVisitors(context, factory))
+        .add(new DataSourceV2RelationVisitor(context, outputFactory))
+        .build();
```

Comment:
This is the cause of your jackson exception

```
JsonMappingException: object is not an instance of declaring class (through reference chain: io.openlineage.client.OpenLineage$RunEvent["inputs"]->java.util.ArrayList[0]->io.openlineage.client.OpenLineage$OutputDataset["namespace"])
```
if you notice, it's complaining about the `OutputDataset`'s namespace, which it reached from the `inputs` field of the `RunEvent`. Jackson is trying to invoke `InputDataset#getNamespace` on an instance of `OutputDataset`. The Since the common visitors are used for both input and output visitors, the factory passed in will reflect the type of dataset that needs to be created by the visitor in the current context. You'll need to use the factory that's been passed in rather than constructing your own `outputFactory`.